rust/kernel/thread.rs

```diff
@@ -0,0 +1,253 @@
+    }
+
+    /// Wakes up the thread.
+    ///
+    /// Note that a newly created thread (e.g. via [`Thread::try_new`]) will not run until a
+    /// [`Thread::wake_up`] is called.
+    ///
+    /// # Context
+    ///
+    /// This function might sleep, don't call in atomic contexts.
+    pub fn wake_up(&self) {
+        // SAFETY:
+        //
+        // `task` is a valid pointer to a kernel thread structure, the refcount of which is
+        // increased in `try_new*`, so it won't point to a freed `task_struct`. And it's not
+        // stopped because `stop` will consume the [`Thread`].
+        unsafe {
+            bindings::wake_up_process(self.task);
+        }
+    }
+
+    /// Stops the thread.
+    ///
+    /// - If the thread hasn't been waken up after creation, the thread closure won't be called,
+    /// and will return `EINTR`. Note that a thread may not be waken up even after
+    /// [`Thread::wake_up`] is called.
+    ///
+    /// - Otherwise, wait for the closure to return or the thread `do_exit` itself.
+    ///
+    /// Consume the [`Thread`] so that it's not accessible. Return the result of the thread
+    /// closure (or the exit code in [`KernelResult`] format).
+    ///
+    /// # Context
+    ///
+    /// This function might sleep, don't call in atomic contexts.
+    pub fn stop(self) -> KernelResult<()> {
+        let ret;
+        // SAFETY:
+        //
+        // `task` is a valid pointer to a kernel thread structure, the refcount of which is
+        // increased in `try_new*`, so it won't point to a freed `task_struct`. And it's not
+        // stopped because `stop` will consume the [`Thread`].
+        unsafe { ret = bindings::kthread_stop(self.task) }
+
+        if ret == 0 {
+            Ok(())
+        } else {
+            Err(Error::from_kernel_errno(ret))
+        }
+    }
+}
+
+impl Drop for Thread {
+    fn drop(&mut self) {
+        // Decreases the refcount of the thread, the thread may still be running after we `drop`
+        // the `Thread`.
+        //
+        // SAFETY:
+        //
+        // At least one refcount is held by `Thread::try_new*` and refcount of `task_struct` is
+        // implemented by atomics.
+        unsafe {
+            rust_helper_put_task_struct(self.task);
+        }
+    }
+}
+
+/// Tries to give up the cpu and let another thread to run.
+///
+/// This maps to kernel's `schedule` function, which is similar to [`std::thread::yield_now`].
+///
+/// # Context
+///
+/// This function might sleep, don't call in atomic contexts.
+///
+/// [`std::thread::yield_now`]: https://doc.rust-lang.org/std/thread/fn.yield_now.html
+pub fn schedule() {
+    // SAFETY:
+    //
+    // If we can schedule back from other thread, then this can be treated as no-ops. A special
+    // case are a thread sets its state to `TASK_DEAD`, and then `schedule` will not come.
+    // Currently we don't have a way to do this safely in Rust, and in the future, we probably
+    // still don't allow it.
```

Comment:
```suggestion
    // SAFETY:
    //
    // If we can schedule back from other thread, then this can be treated as a no-op. A special
    // case occurs when a thread sets its state to `TASK_DEAD`, and then `schedule` will not come.
    // Currently we don't have a way to do this safely in Rust, and in the future, we probably
    // still won't allow it.
```

Also, as with the others: `// SAFETY: ...` instead of adding newlines.